### PR TITLE
C51-257: Custom fields masonry grid and accordions

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview

This PR adds the following changes:

* It changes the way that the masonry grid layout for custom case fields is generated.
* It adds custom stylings for accordions.

⚠️ **Important:** this PR is part of https://github.com/compucorp/uk.co.compucorp.civicaseexrtas/pull/2

## Custom fields masonry grid

The reason for changing the masonry grid structure was because displaying the case outcome is variable and the grid may change depending on whether the outcome panel is displayed or not.

The previous structure had two column elements and would split odd and event elements into the different columns:

<table>
  <tr>
    <td>1</td><td>2</td>
  </tr>
  <tr>
    <td>3</td><td>4</td>
  </tr>
  <tr>
    <td>5</td><td></td>
  </tr>
</table>

This works well when the all the data comes from a single source, but when the data can come from multiple sources this structure can break. This issue can been seen when adding the case outcomes to the list alongside the custom fields:

<table>
  <tr>
    <td>Outcomes</td><td>2</td>
  </tr>
  <tr>
    <td>1</td><td>4</td>
  </tr>
  <tr>
    <td>3</td><td></td>
  </tr>
  <tr>
    <td>5</td><td></td>
  </tr>
</table>

This is how it looks with the current structure:

screen shot 2018-10-19 at 6 44 35 am

Please refer to the previous structure for more details:
https://github.com/compucorp/uk.co.compucorp.civicase/blob/c17aa4c055aa84acc3c6ce52db53a77f3f0f2851/ang/civicase/CaseDetails--tabs--summary--CustomData.html

The solution proposed in this PR is to define all elements inside a single masonry grid directive, and then the directive would take care of splitting the elements into the different columns. A sample usage for the directive is as follows:

```html
<div civicase-masonry-grid>
  <div civicase-masonry-grid-item>Outcomes</div>
  <div civicase-masonry-grid-item>1</div>
  <div civicase-masonry-grid-item>2</div>
  <div civicase-masonry-grid-item>3</div>
  <div civicase-masonry-grid-item>4</div>
  <div civicase-masonry-grid-item>5</div>
</div>
```

The final result looks as follows:
screen shot 2018-10-19 at 6 58 30 am

### Technical details

The masonry directive has a controller that allows child elements to be appended to it in any way they choose:

```js
var vm = this;
vm.$gridItems = [];

vm.addGridItem = function ($gridItem) {
  _.remove(vm.$gridItems, $gridItem);
  vm.$gridItems.push($gridItem);
  $scope.$emit('civicaseMasonryGrid::updated');
};

// Adds the grid item at a specific point in the grid:
vm.addGridItemAt = function ($gridItem, atIndex) {
  _.remove(vm.$gridItems, $gridItem);
  vm.$gridItems.splice(atIndex, 0, $gridItem);
  $scope.$emit('civicaseMasonryGrid::updated');
};

vm.removeGridItem = function ($gridItem) {
  _.remove(vm.$gridItems, $gridItem);
  $scope.$emit('civicaseMasonryGrid::updated');
};
```

The directive itself first appends the two necessary columns, arranges the grid items, and waits for grid item changes in order to update the grid:

```js
(function init () {
  $timeout(function () {
    appendMasonryContainers();
    arrangeGridItems();
    $scope.$on('civicaseMasonryGrid::updated', arrangeGridItems);
  });
})();
```

Creating the containers is done as follows:

```js
function appendMasonryContainers () {
  _.times(2).forEach(function () {
    $('<div></div>')
      .addClass('civicase__summary-tab-tile')
      .addClass('civicase__masonry-grid__container')
      .appendTo($element);
  });
}
```

Arranging the grid items is done as follows:

```js
function arrangeGridItems () {
  $timeout(function () {
    $element.find('[civicase-masonry-grid-item]').detach(); // first removes all the grid elements

    masonryGrid.$gridItems.forEach(function ($gridItem, index) {
      // then moves them to one column or the other:
      var columnIndex = index % 2;
      var $column = $element.find('.civicase__masonry-grid__container').eq(columnIndex);

      $gridItem.appendTo($column);
    });
  });
}
```

By default, the grid items append themselves to the grid on init, and remove themselves when the scope is destroyed, which means the element is also removed from the DOM:

```js
(function init () {
  masonryGrid.addGridItem($element);

  $scope.$on('$destroy', function () {
    masonryGrid.removeGridItem($element);
  });
})();
```

The masonry grid directive uses a simple CSS rule to split the columns:

```scss
.civicase__masonry-grid__container {
  float: left;
  width: 50%;
}
```

## Civicase accordion

A new accordion style was introduced which has the following enhancements as requested by Jon and Shitij:

* The accordion title size is 16px.
* The accordion icon is `>` when closed and `v` when opened.
* The text decoration for the title was removed.
* The padding of the accordion is fixed so it uses 15px.

### Before
pr-before

### After
pr-before

### Technical details

A new `civicase__accordion` scss component is now provided:

```scss
.civicase__accordion {
  .panel-heading {
    padding: 0 0 15px;
  }

  .panel-title {
    font-size: 16px;
  }

  .panel-title a,
  .panel-title a:hover {
    text-decoration: none;
  }

  .panel-title a::before {
    content: '\f105'; // angle-right
  }

  &.panel-open .panel-title a::before {
    content: '\f107'; // angle-down
    margin-left: -4px;
  }

  .panel-body {
    padding: 0 0 15px;
  }
}

```

### CODE REVIEW

**Comment by deb1990** [comment]
@reneolivo please fix the conflicts.

**Comment by AkA84** [comment]
@reneolivo 

> Note that the grid items can be defined inside of a single element, but then they'll be automatically splitted into two columns, where A and C would be on the left column, and B would be on the right column.

why is B placed always on the right column? One would expect that the order of the `civicase-masonry-grid-item` elements  would be respected

Also, what happens if you have 4 `civicase-masonry-grid-item`s? or 5?

**Comment by reneolivo** [comment]
@AkA84 

> Also, what happens if you have 4 civicase-masonry-grid-items? or 5?

When the grid has 4 elements it would look like this:

<table>
  <tr>
    <td>1</td><td>2</td>
  </tr>
  <tr>
    <td>3</td><td>4</td>
  </tr>
</table>


When it has 5 it would look like this:

<table>
  <tr>
    <td>1</td><td>2</td>
  </tr>
  <tr>
    <td>3</td><td>4</td>
  </tr>
  <tr>
    <td>5</td><td></td>
  </tr>
</table>

---

> I'm not sure i understand, does it mean that the directive has an api that I can use to add more elements on runtime?

Yes, they can be added on runtime, removed, or even moved to a different position.

This is useful because the outcomes panel is shown or hidden depending on whether it has information about outcomes. But also, the outcomes need to always be displayed as the first element of the grid.

File: ang/civicase/MasonryGrid.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    var vm = this;
+    vm.$gridItems = [];
+
+    /**
+     * Adds an item element to the grid.
+     *
+     * @param {Object} $gridItem a reference to the item element to be added.
+     */
+    vm.addGridItem = function ($gridItem) {
+      _.remove(vm.$gridItems, $gridItem);
```
Comment (AkA84) [inline ang/civicase/MasonryGrid.js]: you are doing this in basically every method, please put it in a separate function and document it

File: ang/civicase/MasonryGrid.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+     * Masonry grid item link function.
+     *
+     * @param {Object} $scope the directive's scope.
+     * @param {Object} $element a reference to the directive's element.
+     * @param {Object} attrs a map of attributes associated to the element.
+     * @param {Object} masonryGrid a reference to the parent masonry grid directive's controller.
+     */
+    function civicaseMasonryGridItemLink ($scope, $element, attrs, masonryGrid) {
+      (function init () {
+        masonryGrid.addGridItem($element);
```
Comment (AkA84) [inline ang/civicase/MasonryGrid.js]: the logic seems to be in the wrong place...it would make more sense to have this done by `civicaseMasonryGrid`: when the directive initializes, it finds inside itself all the elements marked as `civicase-masonry-grid-item` and  adds them to its controller's list

This directive doesn't really do anything other than doing a job that the parent directive should do, it would be better if you removed it

File: scss/components/_case-custom-fields.scss [hunk 3]
```diff
@@ -1,12 +1,9 @@
 .civicase__case-custom-fields__container {
-  float: left;
-  width: 50%;
-
   &.civicase__summary-tab-tile {
     padding-top: 0;
   }
 
-  .civicase__panel-transparent-header:not(:first-child) {
+  [civicase-masonry-grid-item]:nth-child(1n+2) {
```
Comment (AkA84) [inline scss/components/_case-custom-fields.scss]: this looks pretty ad-hoc, what is it for? why isn't it in `_masonry-grid.scss`?

File: ang/civicase/MasonryGrid.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+          $scope.$on('civicaseMasonryGrid::updated', arrangeGridItems);
+        });
+      })();
+
+      /**
+       * Appends the masonry containers used for splitting the grid items.
+       */
+      function appendMasonryContainers () {
+        _.times(2).forEach(function () {
+          $('<div></div>')
```
Comment (AkA84) [inline ang/civicase/MasonryGrid.js]: honestly i have the feeling that instead of this DOM manipulation, you should have simply used a template for this directive
```html
<!-- directive.html -->
<div class="civicase__summary-tab-tile civicase__masonry-grid__container">
</div>
<div class="civicase__summary-tab-tile civicase__masonry-grid__container">
</div>
```
Then you would add `transclude: true` to the directive
```js
module.directive('civicaseMasonryGrid', function ($timeout) {
  return {
    transclude: true,
    controller: 'civicaseMasonryGridController',
    link: civicaseMasonryGridLink
  };
```
and use [`$transcludeFn`](https://triangular.io/blog/transclusion-transcludefn/) to do all the dom manipulation you need

right now it seems that you're not using the directive fully

File: scss/components/_masonry-grid.scss [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.civicase__masonry-grid__container {
```
Comment (AkA84) [inline scss/components/_masonry-grid.scss]: it would be better to have the parent to apply any layout-related style
```scss
[civicase-masonry-grid] {
  .civicase__masonry-grid__container {
    float: left;
    width: 50%;
  }
}

File: scss/components/_summary-tab.scss [hunk 6]
```diff
@@ -91,6 +97,10 @@
   flex-grow: 1;
 }
 
+.civicase__summary-tab-tile__inner-title {
+  margin-top: 0;
+}
```
Comment (AkA84) [inline scss/components/_summary-tab.scss]: both these two changes in this partial seem very arbitrary, what are they for?

File: ang/civicase/MasonryGrid.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+     * Masonry Grid link function.
+     *
+     * @param {Object} $scope the directive's scope.
+     * @param {Object} $element a reference to the directive's element.
+     * @param {Object} attrs a map of attributes associated to the element.
+     * @param {Object} masonryGrid a reference to the directive's controller.
+     */
+    function civicaseMasonryGridLink ($scope, $element, attrs, masonryGrid) {
+      (function init () {
+        $timeout(function () {
```
[the remark this answers is not in the document]
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: Child elements are not ready until the next tick. The order would be as follows:

* The parent controller is defined.
* The child controller is defined and then adds itself to the parent list.
* the child directive link function runs.
* the parent directive link function runs.

At least the link functions running happen in different ticks, so we need to wait for the child elements to run first.

File: ang/civicase/MasonryGrid.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            .addClass('civicase__masonry-grid__container')
+            .appendTo($element);
+        });
+      }
+
+      /**
+       * Sorts the grid items into one of the two grid containers.
+       */
+      function arrangeGridItems () {
+        $timeout(function () {
```
[the remark this answers is not in the document]
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: I thought it was needed, but I tested it again and it's not. Will remove it.

File: ang/civicase/MasonryGrid.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+          $element.find('[civicase-masonry-grid-item]').detach();
+
+          masonryGrid.$gridItems.forEach(function ($gridItem, index) {
+            var columnIndex = index % 2;
+            var $column = $element.find('.civicase__masonry-grid__container').eq(columnIndex);
+
+            $gridItem.appendTo($column);
```
[the remark this answers is not in the document]
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: > only when all the items are in place, replace the columns in the DOM

replace the columns?

File: ang/civicase/MasonryGrid.js (hunk 2 above)
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: This is how AngularJS recommends child-parent interactions:
https://docs.angularjs.org/guide/directive#creating-directives-that-communicate

Also, when the child has control over when to append itself to the parent, it allows more dynamic control of when to show or hide the element from parent grid.

File: scss/components/_case-custom-fields.scss (hunk 3 above)
Comment (reneolivo) [reply scss/components/_case-custom-fields.scss]: This is a margin that is added to the 2nd element onwards.

<table>
  <tr>
    <td>1</td><td>2</td>
  </tr>
  <tr>
    <td><br>3</td><td><br>4</td>
  </tr>
  <tr>
    <td><br>5</td><td></td>
  </tr>
</table>

This is not done on the grid component because the margin could be different elsewhere.

File: ang/civicase/MasonryGrid.js (hunk 4 above)
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: will investigate more about this!

File: scss/components/_summary-tab.scss (hunk 6 above)
Comment (reneolivo) [reply scss/components/_summary-tab.scss]: I'll remove these changes. They were added before using the new grid and now they are not needed. I forgot to remove them.

File: ang/civicase/MasonryGrid.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (AkA84) [reply ang/civicase/MasonryGrid.js]: or do whatever DOM operation you need to do. the point is, defer as long as possible any DOM operation, and try to have the least amount of them (just 1 would be ideal)

File: ang/civicase/MasonryGrid.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+     * @param {Object} $element a reference to the directive's element.
+     * @param {Object} attrs a map of attributes associated to the element.
+     * @param {Object} masonryGrid a reference to the directive's controller.
+     */
+    function civicaseMasonryGridLink ($scope, $element, attrs, masonryGrid) {
+      (function init () {
+        $timeout(function () {
+          appendMasonryContainers();
+          arrangeGridItems();
+          $scope.$on('civicaseMasonryGrid::updated', arrangeGridItems);
```
[the remark this answers is not in the document]
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: This event is emitted by the controller in order to update the grid (DOM manipulation) when an element changes. How can this be moved to the controller unless we do the actual manipulation on the controller?

File: ang/civicase/MasonryGrid.js (hunk 4 above)
Comment (reneolivo) [reply ang/civicase/MasonryGrid.js]: @AkA84 I tried to work a bit with `$transcludeFn` but I don't think it's useful at all. It gives you a reference to the whole transcluded element instead of individual grid items and they are not even ready by the time I want to manipulate them into different columns.

Comment (AkA84) [reply ang/civicase/MasonryGrid.js]: @reneolivo i really think that you're manually doing something that the directive already does out of the box, but the time constraint is what it is, so let's keep things as they are

File: ang/civicase/MasonryGrid.js (hunk 1 above)
Comment (AkA84) [reply ang/civicase/MasonryGrid.js]: @reneolivo you didn't reply nor act on this comment
